napari/_vispy/overlays/scale_bar.py

```diff
@@ -72,12 +72,37 @@ def _calculate_best_length(self, desired_length: float):
             # return the last index.
             index -= 1
         new_value = PREFERRED_VALUES[index]
+        # validate if new_value is greater than the actual calculated magnitude
+        # to prevent the scale bar to extend beyond the canvas when zooming.
+        # See https://github.com/napari/napari/issues/5914
+        if new_value > new_quantity.magnitude:
+            new_value = new_quantity.magnitude
```

Comment:
I agree with @brisvag I think we want to rethink all that is above, we also want to keep in mind that AFAICT, the `new_quantity` objects can have a value µm or equivalent, so we want to keep the `{:~}` in format.

Reply:
If we return quantity, we can use https://pint.readthedocs.io/en/stable/user/formatting.html to select how we format it and it should automatically format exponent when needs be, so I went overboard in my example.
```
In [11]: f"{one*1e-4:~P}"
Out[11]: '0.0001'

In [12]: f"{one*1e-5:~P}"
Out[12]: '1×10⁻⁵'

In [13]: f"{one*1e-9:~P}"
Out[13]: '1×10⁻⁹'
```